rcl/src/rcl/validate_security_context_name.c

```diff
@@ -0,0 +1,148 @@
+  int * validation_result,
+  size_t * invalid_index)
+{
+  if (!security_context) {
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+  return rmw_validate_namespace_with_size(
+    security_context, strlen(security_context), validation_result, invalid_index);
+}
+
+rcl_ret_t
+rcl_validate_security_context_name_with_size(
+  const char * security_context,
+  size_t security_context_length,
+  int * validation_result,
+  size_t * invalid_index)
+{
+  if (!security_context) {
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+  if (!validation_result) {
+    return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  int t_validation_result;
+  size_t t_invalid_index;
+  rmw_ret_t ret = rmw_validate_namespace_with_size(
+    security_context, security_context_length, &t_validation_result, &t_invalid_index);
+  if (ret != RMW_RET_OK) {
+    return rcl_convert_rmw_ret_to_rcl_ret(ret);
+  }
+
+  if (t_validation_result != RMW_NAMESPACE_VALID &&
+    t_validation_result != RMW_NAMESPACE_INVALID_TOO_LONG)
+  {
+    switch (t_validation_result) {
+      case RMW_NAMESPACE_INVALID_IS_EMPTY_STRING:
+        *validation_result = RCL_CONTEXT_NAME_INVALID_IS_EMPTY_STRING;
+        break;
+      case RMW_NAMESPACE_INVALID_NOT_ABSOLUTE:
+        *validation_result = RCL_CONTEXT_NAME_INVALID_NOT_ABSOLUTE;
+        break;
+      case RMW_NAMESPACE_INVALID_ENDS_WITH_FORWARD_SLASH:
+        *validation_result = RCL_CONTEXT_NAME_INVALID_ENDS_WITH_FORWARD_SLASH;
+        break;
+      case RMW_NAMESPACE_INVALID_CONTAINS_UNALLOWED_CHARACTERS:
+        *validation_result = RCL_CONTEXT_NAME_INVALID_CONTAINS_UNALLOWED_CHARACTERS;
+        break;
+      case RMW_NAMESPACE_INVALID_CONTAINS_REPEATED_FORWARD_SLASH:
+        *validation_result = RCL_CONTEXT_NAME_INVALID_CONTAINS_REPEATED_FORWARD_SLASH;
+        break;
+      case RMW_NAMESPACE_INVALID_NAME_TOKEN_STARTS_WITH_NUMBER:
+        *validation_result = RCL_CONTEXT_NAME_INVALID_NAME_TOKEN_STARTS_WITH_NUMBER;
+        break;
+      default:
+        {
+          char default_err_msg[256];
+          // explicitly not taking return value which is number of bytes written
+          int ret = rcutils_snprintf(
+            default_err_msg, sizeof(default_err_msg),
+            "rcl_validate_security_context_name_with_size(): "
+            "unknown rmw_validate_namespace_with_size() result '%d'",
+            t_validation_result);
+          if (ret < 0) {
+            RCL_SET_ERROR_MSG(
+              "rcl_validate_security_context_name_with_size(): "
+              "rcutils_snprintf() failed");
+          } else {
+            RCL_SET_ERROR_MSG(default_err_msg);
+          }
+        }
+        return RCL_RET_ERROR;
+    }
+    if (invalid_index) {
+      *invalid_index = t_invalid_index;
+    }
+    return RCL_RET_OK;
+  }
+
+  // security_context might be longer that namespace length, check false positives and correct
+  if (t_validation_result == RMW_NAMESPACE_INVALID_TOO_LONG &&
+    security_context_length <= RCL_CONTEXT_NAME_MAX_LENGTH)
```

Comment:
@ivanpauno mind to explain? Also, I have the impression some result is wrong, as this `if` clause is redundant as-is once you get to this point.

Reply:
The security context name can be longer that a namespace.
I'm reusing the namespace check, and in the case the error was that the name was too long, I recheck if that's actually the case. If not, I correct the validation result to valid and return `RCL_RET_OK`.

Note: My logic is valid because the length is the last thing being checked:
https://github.com/ros2/rmw/blob/f76511333709a6f5d1e4543129ec73f2cd2b4f06/rmw/src/validate_full_topic_name.c#L117
https://github.com/ros2/rmw/blob/f76511333709a6f5d1e4543129ec73f2cd2b4f06/rmw/src/validate_namespace.c#L112

Reply:
I may be crazy, but, let's say `tmp_validation_result` equals `RMW_NAMESPACE_INVALID_TOO_LONG` and `security_context_length` exceeds `RCL_SECURITY_CONTEXT_NAME_MAX_LENGTH`. The switch statement [will be skipped](https://github.com/ros2/rcl/pull/515/files#diff-6d1e34c077896ab6d3f8e1a936573cccR65), the  if statement immediately below [will be skipped](https://github.com/ros2/rcl/pull/515/files#diff-6d1e34c077896ab6d3f8e1a936573cccR112) as well, and the implementation will conclude that [the name is valid](https://github.com/ros2/rcl/pull/515/files#diff-6d1e34c077896ab6d3f8e1a936573cccR120).

Reply:
You're right, I failed to get your point the first time.
See https://github.com/ros2/rcl/pull/515/commits/675e4d566319dc784cd514fab12bb2858f34a1f7.